### ISSUE TITLE
`git grep` command hangs forever with high CPU usage

### ISSUE THREAD

**Issue by kaiyoma** [git-for-windows/git#4697]
- [x] I was not able to find an [open](https://github.com/git-for-windows/git/issues?q=is%3Aopen) or [closed](https://github.com/git-for-windows/git/issues?q=is%3Aclosed) issue matching what I'm seeing

### Setup

 - Which version of Git for Windows are you using? Is it 32-bit or 64-bit?

```
$ git --version --build-options

git version 2.42.0.windows.2
cpu: x86_64
built from commit: 2f819d1670fff9a1818f63b6722e9959405378e3
sizeof-long: 4
sizeof-size_t: 8
shell-path: /bin/sh
feature: fsmonitor--daemon
```

 - Which version of Windows are you running? Vista, 7, 8, 10? Is it 32-bit or 64-bit?

```
$ cmd.exe /c ver

Microsoft Windows [Version 10.0.22631.2715]
```

 - What options did you set as part of the installation? Or did you choose the
   defaults?

```
# One of the following:
> type "C:\Program Files\Git\etc\install-options.txt"
> type "C:\Program Files (x86)\Git\etc\install-options.txt"
> type "%USERPROFILE%\AppData\Local\Programs\Git\etc\install-options.txt"
> type "$env:USERPROFILE\AppData\Local\Programs\Git\etc\install-options.txt"
$ cat /etc/install-options.txt

Editor Option: VIM
Custom Editor Path:
Default Branch Option:
Path Option: Cmd
SSH Option: OpenSSH
Tortoise Option: false
CURL Option: OpenSSL
CRLF Option: LFOnly
Bash Terminal Option: MinTTY
Git Pull Behavior Option: Merge
Use Credential Manager: Enabled
Performance Tweaks FSCache: Enabled
Enable Symlinks: Enabled
Enable Pseudo Console Support: Disabled
Enable FSMonitor: Disabled
```

 - Any other interesting things about your environment that might be related
   to the issue you're seeing?

No

### Details

 - Which terminal/shell are you running Git from? e.g Bash/CMD/PowerShell/other

Bash

 - What commands did you run to trigger this issue? If you can provide a
   [Minimal, Complete, and Verifiable example](http://stackoverflow.com/help/mcve)
   this will help us understand the issue.

```
git grep "ace.*editor"
```
 - What did you expect to occur after running these commands?

The command should work as normal (print search results).

 - What actually happened instead?

Running `git grep "ace.*editor"` in our repo is causing git to hang forever
with very high CPU usage. After running the command, I immediately get a
few results, then it hangs forever with this message:

```
Waiting for data... (^X or interrupt to abort)
```

CPU usage remains high and the command never completes.

There is something about this specific command that is causing git to
misbehave. If I change the strings slightly (e.g. `git grep "monaco.*editor"`)
then the command completes as usual.

I tried adding a `--verbose` flag to the command to get more info, but
that doesn't seem to work. I'm happy to provide more info if you like, but I
can't provide the repo as a whole (it's a private repo).

 - If the problem was occurring with a specific repository, can you provide the
   URL to that repository to help us with testing?

Unfortunately I cannot. This is a large, private repo.

**Comment by dscho** [comment]
> This is a large, private repo.

My guess is that this, combined with probably a large number of `ace` matches (or matches in really long lines) may be the culprit. `git grep`'s performance depends on the specificity of the provided regular expression, after all.

**Comment by kaiyoma** [comment]
> My guess is that this, combined with probably a large number of `ace` matches (or matches in really long lines) may be the culprit. `git grep`'s performance depends on the specificity of the provided regular expression, after all.

`git grep` commands in this repo normally take only a second or so. But here's something very interesting. If I make the regex _broader_, it actually completes!

```
$ time git grep "ace.*editor" > /dev/null

(this command has to be killed because it hangs forever)

$ time git grep "ace.*ed" > /dev/null

real    0m1.763s
user    0m0.000s
sys     0m0.016s
```

It seems like we're tickling some kind of corner case with this exact command.

**Comment by dscho** [comment]
Right. Could you experiment with the `-P` option (i.e. using PCRE2 instead of POSIX' regexp functionality)? Maybe that helps, too.

**Comment by kaiyoma** [comment]
> Right. Could you experiment with the `-P` option (i.e. using PCRE2 instead of POSIX' regexp functionality)? Maybe that helps, too.

Using `-P` fixes things:

```
$ time git grep -P "ace.*editor"
(redacted)

real    0m1.115s
user    0m0.000s
sys     0m0.000s
```

**Comment by dscho** [comment]
Okay, thank you for testing. My suspicion is that [the regex code we carry in `compat/regex/`](https://github.com/git-for-windows/git/tree/HEAD/compat/regex) may need updating: It is [vendored in from `gawk`](https://github.com/git-for-windows/git/commit/d18f76dccff1fc56221846926edb4542d88f60e0) and has last been updated in 2010 (!).

Looking around a bit, it seems that [`gawk` switched to gnulib's version of regex in 2018](https://git.savannah.gnu.org/cgit/gawk.git/commit/?id=b98257919b20bdfc14f363761cc6215c1ad8bcee). Which suggests that we should probably try to synchronize the relevant files in `compat/regex/` with [gnulib](https://git.savannah.gnu.org/cgit/gnulib.git/tree/lib/)'s versions (and then make sure that any [changes we added on top of `gawk`](https://github.com/git-for-windows/git/commits/HEAD/compat/regex) are replayed).

@kaiyoma want to give it a try?

**Comment by Chand-ra** [comment]
@dscho may I have a go at this?

**Comment by dscho** [comment]
> @dscho may I have a go at this?

No need to ask for permission :-)

**Comment by Chand-ra** [comment]
Hey @dscho, I was working on this issue and am basically done with replacing the files and re-applying the commits but when trying to compile the resulting files, the compiler spits a bunch of header-file related errors like:
>'size_t' is defined in header '<stddef.h>'; did you forget to '#include <stddef.h>'?

but <stddef.h> is already #included in regex.h. I have added a [draft PR](https://github.com/gitgitgadget/git/pull/1641) for the work upto now. Any idea on how to resolve this?

**Comment by rimrul** [comment]
From a cursory look all the CI jobs seem to be upset about the `#include <libc-config.h>` in `compat/regex/regex.c`.  That's probably there for `__GNUC_PREREQ()`. You can probably use `GIT_GNUC_PREREQ()` from `git-compat-util.h` instead.
